Edge device creation, configuration, and device bundles

c-ryan-k opened the pull request:
Adds `az iot edge devices create` command that allows users to create and configure multiple edge device scenarios (and associated device configuration settings) using a configuration file (YAML or JSON) or inline device arguments.

The command performs the following actions after processing input:
- Verify parent-child relationships between devices
- [Optional] Delete all devices on the hub before creation
- Create all new device-identities
- Assign parent scopes to devices
- Set deployments / modules on devices
- Creates device config bundle (.tgz) that contains:
  - Device certificates
  - Root CA certificate (self-signed or imported)
  - Install script and readme


---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

c-ryan-k committed:
Initial nested edge hierarchy implementation

c-ryan-k committed:
Moved edge hierarchy command into provider pattern

c-ryan-k committed:
Remove Default Edge Agent from edge hierarchy config

c-ryan-k committed:
Linter fixes

c-ryan-k committed:
Pattern updates and initial tests

c-ryan-k committed:
Cleanup / linting / refactoring, test updates

c-ryan-k committed:
Added integration tests and minor fix in multi-step hierarchy logic

c-ryan-k committed:
Help Updates

Added help and examples for edge hierarchy commands

Modified inline device args to parse 'id' instead of 'device_id' for brevity, updated tests

c-ryan-k committed:
Added deployment config processing and unit tests

c-ryan-k committed:
Updated YAML parsing exception handling

c-ryan-k committed:
Merge branch 'azure-dev' into nested_edge_rebase

c-ryan-k committed:
Cleanup

reverted formatting changes to hub.py

removed comment from int test file

c-ryan-k committed:
Certificate creation for nested edge devices

c-ryan-k committed:
Device config / bundle updates, minor fixes

c-ryan-k committed:
Refactoring / fixes / device bundle improvements

c-ryan-k committed:
Added test files for e2e tests

c-ryan-k committed:
Attempted to fix CA/CSR cert signing, notes for necessary unit tests

c-ryan-k committed:
File operation tests

c-ryan-k committed:
Some cert and file  refactoring

c-ryan-k committed:
Configuration script utility tests

c-ryan-k committed:
Updated tests

c-ryan-k committed:
Refactoring nested edge logic to be a bit cleaner, updated test

c-ryan-k committed:
Merge branch 'dev' into nested_edge_rebase

c-ryan-k committed:
Wave of PR feedback

c-ryan-k committed:
Help / param updates

c-ryan-k committed:
Updates / refactoring / tests

c-ryan-k committed:
Testing updates

Certificate whitespace fixes

added all --device arg config values

Updated unit tests

Added device bundle checks to integration tests

c-ryan-k committed:
Refactoring and linting

Moved new tuples / classes to shared

Fixed linter issues and styling

c-ryan-k commented:
I realized there's going to be some line-length linter errors for the (quite verbose) scripts I threw in `constants` instead of copying the files from extension source to disk.

Will investigate a better solution / place for these.

digimaun commented on azext_iot/common/shared.py:
```diff
@@ -304,5 +309,43 @@ class SHAHashVersions(Enum):
     SHA256 = 256
+
+
+class CertInfo(TypedDict):
```
We currently support Python 3.7 ...TypedDict was introduced in Python 3.8

vilit1 replied:
one option is to say we only support 3.8 now

digimaun replied:
Disagree, we should deprecate python versions when they aren't supported anymore. Python 3.7 EOL is in 8 months.

c-ryan-k replied:
The beauty of TypedDict is that we can just replace it with `Dict[str, str]` or `Dict[str, Any]` - it was just a convenience type for the certificate object format we pass around. I'll get rid of the TypedDict for now.

digimaun replied:
Makes sense. I'm all for using the best facilities of Python we can, but we have to balance with compatibility and casting the widest (safe/secure) net of usage. We saw offline there are **still** notable usage of legacy/out-of-support Py versions.

c-ryan-k replied:
Yup - totally agree. I actually didn't see your comment when I typed mine, so I was just reiterating support for waiting to use it until we're able to ensure compatibility. If anything, it was worth using during dev for the intellisense alone 😆

digimaun commented on azext_iot/constants.py:
```diff
@@ -54,3 +54,127 @@
 IOTDPS_TRACK_2_SDK_MIN_VERSION = "1.0.0"
+
+# Edge device TOML default values
+DEVICE_CONFIG_TOML = {
```
Can we put these edge config and script related attributes in a separate edge oriented module?

c-ryan-k replied:
I noticed that in another PR we have a new `common.py` file created in `azext_iot/iothub` - do you think this is the right spot for these, or should we:
- Add a `common` folder with (examples):
  - `hub.py`
  - `devices.py`
  - `edge.py`
 - Create a new `edge_common.py` strictly for edge-related common code.
 We could also move some of the `device_identity.py` provider functions into this file , like `_process_nested_edge_config_content` or `create_edge_device_config`

digimaun replied:
What is different about these newly introduced elements is they are focused around device side scripting & configuration which IMO does not harmonize with the organization of common modules in iothub, deviceupdate, digitaltwins, central etc -- those are in the realm of enums (often times to align service & client options), resource ids, naming functions, generic messages (arguably can be moved) and namespace specific client settings.

More prefer the second option, but for these specific attributes, something like `edge_device_config.py` seems appropriate because just reading the module name I have an idea it will be edge device configuration centric where `edge_common` can be ambiguous with respect to what part of edge overall (service side?) + what I was saying earlier where it wouldn't be organized like our other namespace specific common modules.  I also do like the thought of moving the edge config specific functions/classes into an edge config based module rather than mixing with device_identity.

c-ryan-k commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,38 @@ def load_iothub_help():
+        - name: Create a simple nested edge device hierarchy (deleting existing devices)
+            with an existing root CA, using x509 auth, and specify a custom device bundle output path.
+          text: >
+            az iot edge hierarchy create -n {hub_name} --clean --root-cert "root_cert.pem" --root-key "root_key.pem" --device-auth x509_ca
```
I need to add a help example for extra `--device` arguments (hostname, edge_agent, container_auth)

c-ryan-k commented on azext_iot/iothub/commands_device_identity.py:
```diff
@@ -0,0 +1,68 @@
+    )
+
+
+def iot_delete_devices(
```
May remove this provider method until we decide on a suitable namespace for the command, like `az iot hub device-identity clean/delete --all` etc.

digimaun replied:
Since it takes a list of device Id's (its not always all devices) I could see it being useful now, as in `az iot hub device-identity delete` today takes a single deviceId , we could flex it to take 1 or more values. Someone could delete 1 or 1000 in a single command execution. Have to think about the etag though. 2 immediate thoughts are to keep an optional etag single value like now (where it would be useful for a single device delete), or also flex etag to take 1 or more values, where the etag value index corresponds to the device id value index. If the length of values mismatch raise an error.

c-ryan-k replied:
Hmm..we could use `eTag` as `*` in this command for the user (letting users know it won't check eTags, of course) or maybe support eTags as another param like `az iot hub device-identity delete --device id=test eTag=_etag_`

digimaun replied:
For the first idea, right now for the device identity delete command, if the user does not provide an etag then * will be assumed - are you thinking if multiple device id's are provided, then the only choice is '*' + let users know of that etag behavior change?

For the second one, an initial though is it _could_ be confusing have both `--device-id` and `--device k=v` styles but we we can simmer on it.

We don't need to make a decision in this PR.

c-ryan-k replied:
Have a note to follow-up on the implementation of this command to allow users to delete multiple device-identities at once.

c-ryan-k commented on azext_iot/tests/iothub/devices/hierarchy_configs/fake_edge_container_auth.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Add to credscan suppressions / document mock creds to avoid pipeline failures

c-ryan-k commented on test.yml:
```diff
@@ -0,0 +1,31 @@
+config_version: "1.0"
```
Will remove this file for merge

c-ryan-k commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,658 @@
+        for check, err in [
+            (version, "No schema version specified in configuration file"),
+            (hub_config, "No `iothub` properties specified in configuration file")
+            # (len(devices_config), "No devices specified in configuration file")
```
We should probably throw an exception if a config file is used without any devices - going to uncomment this and update tests.

vilit1 commented on .gitignore:
```diff
@@ -289,8 +289,8 @@ __pycache__/
 
 # Virtual environment
 env/
-env27/
```
has python 2 not been removed?

c-ryan-k replied:
Not in our .gitignore, at least - I can remove the newly added line if we want, just figured maybe there's someone out there with a 2.x env folder floating around :)

c-ryan-k committed:
Removes TypedDict usage

c-ryan-k committed:
Credscan fixes

c-ryan-k committed:
Refactored edge_device_config members into separate module

c-ryan-k committed:
Help and parameter updates

c-ryan-k committed:
Various minor tweaks

Renamed edge hierarchy test file

Throw an error if nested edge config file has no devices

c-ryan-k committed:
Delete test.yml

c-ryan-k committed:
Delete extra test config files

c-ryan-k committed:
Certificate updates

Removed CSR from signed certificate creation

Added support for private key passwords

c-ryan-k committed:
Parameter updates

Adds default edge_agent and device_config_template (toml) path arguments to hierarchy create

Moves --device parameter processing and nArgs functions to edge utilities

Added unit tests for parameter processing

digimaun commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,81 @@ def load_iothub_arguments(self, _):
+            ],
+            help="Directory path to output device configuration bundles.",
+        )
+        context.argument(
```
Any opportunities for organizing related arguments as arg groups? For example an arg group for cert related args.

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,43 @@ def load_iothub_help():
         short-summary: Upload a local file as a device to a pre-configured blob storage container.
     """
+
+    helps["iot edge hierarchy"] = """
```
We discussed the command name change making sure there is ample examples of hierarchy based + non-hierarchy based edge device configuration examples.

digimaun commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,409 @@
+        )
+
+        self.hub_resolver = SdkResolver(target=self.target)
+        self.service_sdk = self.hub_resolver.get_sdk(SdkType.service_sdk)
```
SdkResolver is already available in the base IoTHubProvider class, you can use the base method `get_sdk()`

digimaun commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,409 @@
+        self,
+        devices: Optional[List[List[str]]] = None,
+        config_file: Optional[str] = None,
+        visualize: Optional[bool] = False,
```
nit, because the default value False is assigned, I don't see the semantic value that Optional provides for the bool.

digimaun commented on azext_iot/common/utility.py:
```diff
@@ -146,6 +150,44 @@ def process_json_arg(content: str, argument_name: str = "content", preserve_orde
+        )
+
+    try:
+        import toml
```
Where is the `toml` dependency coming from? If from az cli which part of it?

c-ryan-k replied:
I could have sworn it was coming from azure-cli but I think I just realized it's coming from `azdev` - might need to add this to setup / etc

c-ryan-k replied:
After some brief searching, I may look into `tomli/tomli-w` as well, I was only using `toml` since I figured it was already in our dependency list

digimaun replied:
Sounds good `tomli[-w]` looks cool

digimaun commented on azext_iot/common/utility.py:
```diff
@@ -146,6 +150,44 @@ def process_json_arg(content: str, argument_name: str = "content", preserve_orde
         )
 
 
+def process_yaml_arg(path: str) -> dict:
```
nit, might as well use a Dict generic for return type like in the process toml function

digimaun commented on azext_iot/common/utility.py:
```diff
@@ -146,6 +150,44 @@ def process_json_arg(content: str, argument_name: str = "content", preserve_orde
+        )
+
+
+def process_toml_content(path: str) -> Dict[str, Any]:
```
nit, do you think ending in `_content` is appropriate? Just curious since we have process_yaml_**arg** and process_json_**arg**.

c-ryan-k replied:
I can change it to `_arg` - I think I chose this initially since it was a bit different than JSON (which can be passed in as a string or a path) - though YAML is processed the same way so I just missed one of these 🙃

digimaun commented on azext_iot/common/shared.py:
```diff
@@ -304,5 +309,33 @@ class SHAHashVersions(Enum):
+
+
+# Utility classes for nested edge config file values and device arguments
+class EdgeContainerAuth(NamedTuple):
```
Is a EdgeContainerAuth valuable if no attributes are set? Typing shows nothing is required (everything optional).

c-ryan-k replied:
Sorry that was a bit of laziness left over from some object prototyping, I *believe* all of these are required, but there's a chance that user/pass may be optional - will investigate and update.

digimaun commented on azext_iot/common/fileops.py:
```diff
@@ -0,0 +1,47 @@
+    target_directory: str,
+    tarfile_path: str,
+    tarfile_name: str,
+    overwrite: Optional[bool] = False,
```
Another place (and above function) where I think we can remove Optional.

digimaun commented on azext_iot/tests/iothub/devices/hierarchy_configs/invalid/duplicate_device_config.yml:
```diff
@@ -0,0 +1,16 @@
+config_version: "1.0"
+# We don't need iothub info from the config schema
```
We talked about exposing one or more markdown document at least in our project docs that shows or links to example(s) of the config file format and describes key usage information.

To do this, we should concretely pin the structure of the format and version it. Best practice conventions should be applied, and because the CLI supports either json or yaml, I don't think its a bad idea to align on camelCase since that is an accepted practice for both json and yaml.

digimaun commented on azext_iot/common/utility.py:
```diff
@@ -146,6 +150,44 @@ def process_json_arg(content: str, argument_name: str = "content", preserve_orde
+    try:
+        import toml
+
+        with open(path, "rt", encoding="utf-8") as f:
```
Think "r" is a synonym for "rt" but its good to be explicit.

digimaun commented on azext_iot/common/shared.py:
```diff
@@ -304,5 +309,33 @@ class SHAHashVersions(Enum):
+
+
+class NestedEdgeDeviceConfig(NamedTuple):
+    device_id: str
```
The minimum operation with just a device Id is to create an edge enabled device, is that right?

c-ryan-k replied:
Yes, correct, technically that is the only property you need in the config to create devices.
For this scenario, they're all assumed edge devices, but this could easily be `DeviceConfig` with a property of `edge: bool`

digimaun commented on azext_iot/common/shared.py:
```diff
@@ -304,5 +309,33 @@ class SHAHashVersions(Enum):
     SHA256 = 256
+
+
+# Utility classes for nested edge config file values and device arguments
```
I think these utility namedtuples should be located within the `iothub` namespace. This module is legacy and is in large part supporting the classic operations/* implementation style. Placing them either in `edge_device_config` or a new `.../iothub/common.py` would be more optimal.

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,198 @@ def open_certificate(certificate_path: str) -> str:
     return certificate.rstrip()
+
+
+# TODO - Unit test, compare with test_utils::_generate_root_certificate
```
Tiny note to remove TODOs **if** they have been satisfied for less future confusion.

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,198 @@ def open_certificate(certificate_path: str) -> str:
+# TODO - Unit test, compare with test_utils::_generate_root_certificate
+def create_root_certificate(
+    subject: Optional[str] = "Azure_IoT_CLI_Extension_Cert",
+    valid_days: Optional[int] = 365,
```
Is `Optional` appropriate here? What does `valid_days = None` mean?

c-ryan-k replied:
Like you've probably seen elsewhere in my usage of `Optional`, this is likely my Typescript brain fighting with my Python brain :)

I'll make another pass at updating method signatures and object/tuple definitions.

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,198 @@ def open_certificate(certificate_path: str) -> str:
+
+# TODO - Unit test, compare with test_utils::_generate_root_certificate
+def create_root_certificate(
+    subject: Optional[str] = "Azure_IoT_CLI_Extension_Cert",
```
A subject str is always required right? We should only use `Optional` if `None` is allowed.

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,198 @@ def open_certificate(certificate_path: str) -> str:
+def create_root_certificate(
+    subject: Optional[str] = "Azure_IoT_CLI_Extension_Cert",
+    valid_days: Optional[int] = 365,
+    key_size: Optional[int] = 4096,
```
optional key_size makes sense if say the parameter was assigned a value of None, then if falsey, the function would assign  key_size the value of 4096.

digimaun commented on azext_iot/iothub/edge_device_config.py:
```diff
@@ -0,0 +1,469 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
```
Seems appropriate to put this module in providers/, thoughts?

c-ryan-k replied:
Now that I've finished a bunch of other renaming / changes - I'm still thinking about this.

Right now the providers folder only contains actual provider modules (`base`, `device_identity`, `device_messaging`, etc.).

Since this code is used almost exclusively by the new `device_identity` provider, but isn't a provider itself, it seems odd to just throw it in that folder... would it make sense to move some providers into folders (`providers/device_identity/device_identity_provider.py`) in case we extend them? Or maybe just another folder inside `providers` - `providers/[utility | common | helpers]`?

digimaun replied:
The first option (`providers/device_identity/device_identity_provider.py`) is not my favorite because it looks too nested and repeated.

`from iothub.providers.device_identity.device_identity_provider` ...I know I'm importing an iothub provider from the first 2 segments.

And IMO we don't have enough independent modules around device identity at this stage to create a new nested directory.

Then `common` is overloaded -  I'm good with putting it in a new helpers directory nested under providers or even as a sibling to providers i.e.  ........`iothub/providers/helpers/edge_device_config.py` or `iothub/helpers/edge_device_config.py`

digimaun commented on azext_iot/iothub/edge_device_config.py:
```diff
@@ -0,0 +1,469 @@
+EDGE_CONFIG_SCRIPT_PARENT_HOSTNAME = """
+# ======================= Set Parent Hostname =======================================
+
+read -p "Enter the parent hostname to use: " parent_hostname
```
What about ensuring a device or its parent hostname is known? As in are `EDGE_CONFIG_SCRIPT_PARENT_HOSTNAME` and `EDGE_CONFIG_SCRIPT_HOSTNAME` must haves?

digimaun commented on azext_iot/iothub/edge_device_config.py:
```diff
@@ -0,0 +1,469 @@
+)
+
+EDGE_LINUX_TUTORIAL_LINK = (
+    "https://learn.microsoft.com/en-us/azure/iot-edge/how-to-provision-single-device-linux-symmetric"
```
I wonder if these links should be aka.ms ones, unless they will stay the same forever.

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,43 @@ def load_iothub_help():
+        type: command
+        short-summary: |
+          Deploy nested edge hierarchies to an IoT Hub.
+          This can be done with inline device arguments or by supplying a nested edge configuration file in YAML or JSON format.
```
`long-summary` can be used where there is ample enough information to provide in help. Short-summary for each command will be shown in its entirety when showing help for a command group.

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,43 @@ def load_iothub_help():
+            containing 2 parent devices with 1 child device each. Also specifies module deployments
+            to configure the devices and visualizes the operations as they progress.
+          text: |
+            az iot edge hierarchy create -n {hub_name} --clean --visualize
```
General question, currently the device bundle output is always produced right? If so, should there be an opt-in way to activate, or a way to deactivate? 

Scenario I'm thinking - what if a user wants to create or update a set of edge device identities but don't need the device side setup scripts?

c-ryan-k replied:
I agree, this was something that I realized as well - we should either:
- Specify that the device bundle output always happens unless a `--no-output`-esque param is provided
- Only create device bundle output if a path is provided.

c-ryan-k committed:
Fixed parent hostname processing from inline device args

c-ryan-k committed:
Wave of PR feedback

Rename 'edge hierarchy' to 'edge devices'

Configured edge devices create to only output files if path is provided

Update help and examples

Clean up use of [Optional] types

Moved edge devices tuples into iothub.common

Updated edge links to use aka.ms short-links

c-ryan-k committed:
Second wave of updates

Added override values from command line to edge config file processing function

Changed device_auth_type signature to string

Changed from x509_ca to x509_thumbprint for self-signed auth

Updated SHA version on self-signed-cert so thumbprints match

c-ryan-k committed:
Added tomli and tomli-w to setup, thirdpartnotices

c-ryan-k committed:
Linting and style fixes

c-ryan-k committed:
Oops

c-ryan-k committed:
Replaced old usage of toml with tomli

c-ryan-k committed:
Snap to schema

Updated snake casing in configs to camelCasing

Added schema validation to config file parsing

Updated error handling

c-ryan-k committed:
Various minor tweaks

Add print statement to show device bundle output information if necessary

Updated help and params

Added a few links to sample devices config file

c-ryan-k committed:
Merge branch 'dev' into nested_edge_device_bundle

c-ryan-k committed:
Testing updates

c-ryan-k committed:
Refactor of edge_device_config into iothub/providers/helpers

c-ryan-k committed:
Merge branch 'dev' into nested_edge_device_bundle

c-ryan-k committed:
Updates to device hub-auth certificate generation, minor refactoring

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
     return certificate.rstrip()
+
+
+def create_root_certificate(
```
can we combine this with create_self_signed_certificate and/or replace it? There is some customization that would need to be added (such as file prefix, etc)

It may also be helpful to add docstrings here as to what this will be used for (ex: create a root certificate that will be uploaded to the iot hub)

c-ryan-k replied:
I've added docstrings to the certificate commands. For the time being, my recommendation would be to leave these as separate calls, as we may iterate on some of the Edge device certificate generation, and until we know we have one function to rule them all (certs) I think we should leave them separate just in case.

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    }
+
+
+def create_signed_cert(
```
It may also be helpful to add docstrings here as to what this will be used for (ex: create a device certificate signed by a root certificate in the iot hub and use that for device registration)

vilit1 replied:
also this looks like something that can replace the other signed cert function

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption(),
+    ).decode("utf-8")
```
are we not dumping/writing root certificate and key?

c-ryan-k replied:
For these commands (since we create a lot of certs) I made a separate `fileops` command to`write_content_to_file`, so we can do it separately, doesn't necessarily need to be an overload on this command.

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    valid_days: Optional[int] = 365,
+) -> Dict[str, str]:
+
+    private_key = rsa.generate_private_key(public_exponent=65537, key_size=4096)
```
how come we can customize key_size for root but not here? Do we need to be able to customize key_size?

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    for path in [cert_path, key_path]:
+        if not exists(path):
+            raise FileOperationError(
+                "Error loading certificates. " f"No file found at path '{path}'"
```
why not one string

c-ryan-k replied:
Not sure, this might have just been an autoformat / linting miss.

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    subject: str = "Azure_IoT_CLI_Extension_Cert",
+    valid_days: int = 365,
+    key_size: int = 4096,
+) -> Dict[str, str]:
```
do we want to be able to create certs with passwords? (this applies to the function after this too)

c-ryan-k replied:
I don't think so. I think that's best left to the users to create their own certs / keys with passwords, and then they can BYOK with the `--root-cert`, `--root-key`, and `root-pass` parameters.

I think it's a possibility and a nice-to-have, but the way the logic works now, we'd end up overloading `--root-pass` with a "set root password" if creating a self-signed cert, or "use root password" if using an existing protected key.

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    }
+
+
+def load_ca_cert_info(
```
add a short docstring to describe this vs open_certificate uses

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    output_dir: Optional[str] = None,
+    output_file: Optional[str] = None,
+) -> str:
+    cert_content = "".join(certs)
```
I am assuming this works cause I do not remember if the certs need to have a newline between all of them...

also interesting that you have this separate from creating signed certificates - even though it does make the signed certificates easier to understand cause one thing is being done, usually the signed certificates (specifically device) need a cert chain...

c-ryan-k replied:
Yeah, this is pretty much just a helper method for `"".join(any_string_arr) -> write_content_to_file` - it just so happened it was explicitly useful for writing certificate files so it takes a `list[cert_str]` and writes them all to a single file.

vilit1 commented on azext_iot/common/certops.py:
```diff
@@ -125,3 +134,201 @@ def open_certificate(certificate_path: str) -> str:
+    certs: List[str],
+    output_dir: Optional[str] = None,
+    output_file: Optional[str] = None,
+) -> str:
```
for things as scary as certs, it is safer to have a short docstring with what this is doing and where it will be used

vilit1 commented on azext_iot/common/fileops.py:
```diff
@@ -0,0 +1,47 @@
+        f.write(write_content)
+
+
+def tar_directory(
```
why do we need tar files...

c-ryan-k replied:
The end result of `edge devices create` if an `--output` is provided is both the Hub configuration changes (create device, set parents, set modules, etc) as well as a device bundle to install on edge devices themselves, to setup local edge agent configs and certificate paths.

c-ryan-k replied:
So since most edge use-cases are on linux endpoints/devices - we decided to use tar to compress the device bundles.

vilit1 commented on azext_iot/common/utility.py:
```diff
@@ -146,6 +150,44 @@ def process_json_arg(content: str, argument_name: str = "content", preserve_orde
+            return yaml.load(f, Loader=yaml.SafeLoader)
+    except Exception as ex:
+        raise InvalidArgumentValueError(
+            "Failed to parse yaml for file '{}' with exception:\n{}".format(path, ex)
```
YAML?

Also we could make these errors into constants since these and the corresponding toml ones are mostly the same; this will be helpful if we decide to support other file types too

c-ryan-k replied:
Not sure if this is a question on the capitalization (or lack thereof) of `YAML` or a question on the usage of it as a file format 😆

vilit1 replied:
it is two things:
1. the lack of capitalization 
2. the error messages are repeats

I can also question why YAML and then get the wrath of people who like it more than json

vilit1 commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,50 @@ def load_iothub_help():
+
+    helps["iot edge devices"] = """
+        type: group
+        short-summary: Commands to create and manage IoT Edge devices.
```
right now there is only create - is this expected?
Can this summary just say manage (without the create)

c-ryan-k replied:
Correct, this simply adds the ability to create groups of new edge devices on a hub (and at the same time clean the hub if necessary). Can definitely change to manage, felt like create was the star here so hence it's inclusion :)

vilit1 commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,50 @@ def load_iothub_help():
+
+    helps["iot edge devices create"] = """
+        type: command
+        short-summary: Create configuratations of edge devices in an IoT Hub.
```
creating configurations of edge devices != creating edge devices, right?

c-ryan-k replied:
Yeah I can see how this could be confused with "Edge Configurations" but this is creating... (not hierarchies, but... topologies?)

I need to think of a better word than "collection", "group", or "a bunch of"

vilit1 replied:
Why is "Create edge devices in an IoT Hub" not good? Otherwise collection would be ok or even multiple devices. Group can be confused with device groups

vilit1 replied:
Why is "Create edge devices in an IoT Hub" not good? Otherwise collection would be ok or even multiple devices (instead of a bunch of). Group can be confused with device groups...

vilit1 commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,88 @@ def load_iothub_arguments(self, _):
+            options_list=["--device", "-d"],
+            nargs="+",
+            action="append",
+            help="Space-separated key=value pairs corresponding to properties of the edge device to create. "
```
what is the max amount of devices a person can add at a time?

c-ryan-k replied:
There is technically no defined maximum, you are only limited by input configuration / shell / how long your device arguments are, etc.

vilit1 commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,50 @@ def load_iothub_help():
+            Also specifies output path for device certificate bundles.
+          text: |
+            az iot edge devices create -n {hub_name} --out {device_bundle_path}
+            --device id=parent_1 deployment=/path/to/parentDeployment_1.json
```
we may need a warning in help (long summary?) for a person trying to add too many devices and hitting the command line input limit

vilit1 replied:
actually the shell will prob handle it and it should be pretty obvious if we get an issue about this. This can be ignored then

vilit1 commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,50 @@ def load_iothub_help():
+            az iot edge devices create -n {hub_name} --device-auth x509_thumbprint --default-edge-agent {default_edge_agent}
+            --device id=device_1 hostname={FQDN}
+            --device id=device_2 edge_agent={custom_agent}
+            --device id=parent hostname={FQDN} edge_agent={custom_agent} container_auth={path_or_json_string}
```
I hate commands that have key=value args but this is probably for the best here

digimaun replied:
Hate is a strong word, why so? Assuming key value pairs are well defined and there are examples -  there is flexibility being able to support both inline and from file device definition.

vilit1 replied:
talked about this - the solution is better help

vilit1 commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,50 @@ def load_iothub_help():
+            --device id=parent_2 deployment=/path/to/parentDeployment_2.json
+            --device id=child_2 parent=parent_2 deployment=/path/to/child_deployment_2.json
+
+        - name: Create a simple nested edge device configuration with an existing root CA, using x509 auth, and specify a custom device bundle output path.
```
Is there a way to do this while generating a root CA for the hub? (Assuming that the user does not have the root ca)

What if the user has the device certs and wants to use those?

c-ryan-k replied:
We need to be careful not to confuse the Edge device root cert (which is different than a certificate stored on the hub) with the hub root CA - the certificates created by this command are purely for either: edge-to-edge communication (all signed by the same self-signed or imported root certificate) or for edge-to-hub device identity certificates.

Regarding bring-your-own device certs, currently you can regenerate all the device certs based off of an existing root cert you import with the command, but there is no mechanism to individually specify existing certificates per-device.

vilit1 replied:
certs are confusing

vilit1 commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,50 @@ def load_iothub_help():
+            --device id=device_2 edge_agent={custom_agent}
+            --device id=parent hostname={FQDN} edge_agent={custom_agent} container_auth={path_or_json_string}
+
+        - name: Delete all existing devices on a hub and create new devices based on a configuration file (with progress bars and visualization).
```
cant we just tell the user "run the delete command first" (this will require creating said delete command) rather than slapping this into create and making it have more switches/toggles?

c-ryan-k replied:
For the sake of this command, since it's being used to create specifically designated device scenarios, I would argue it's acceptable to have a `--clean` switch here. That's not to say we shouldn't also eventually have a `devices clean` command, but I don't think adding one extra switch to a pretty hefty "infra as config" type command complicates it too much

vilit1 commented on azext_iot/iothub/common.py:
```diff
@@ -0,0 +1,34 @@
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+# Utility classes for edge device configs
```
move this comment to the actual named tuples

also it would be nice to make a named tuple for certificate props that we care about (tho it can be a future task)

vilit1 commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,88 @@ def load_iothub_arguments(self, _):
+            action="append",
+            help="Space-separated key=value pairs corresponding to properties of the edge device to create. "
+            "The following key values are supported: `id`, `deployment` (inline json or path to file), `hostname`, "
+            "`parent`, `edge_agent`, and `container_auth` (inline json or path to file). "
```
is there a reference with respect to what each of these key values mean?


Here is a location help from `az cosmosdb create` that may help with formatting
![image](https://user-images.githubusercontent.com/73560279/206558123-dd2f5958-0440-47c1-8d7a-883cd199d66a.png)

vilit1 replied:
will look for a better example

c-ryan-k replied:
These are the same values that are in the configuration yaml/json schema per-device, so perhaps we could just add a note near the link to that sample file that declares that they are the same properties / types as in that schema.

vilit1 replied:
as long as a clueless person can figure out the typing I am good

vilit1 commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,88 @@ def load_iothub_arguments(self, _):
+            "clean",
+            options_list=["--clean", "-c"],
+            arg_type=get_three_state_flag(),
+            help="Deletes all devices in target hub before creating new devices.",
```
this sounds like a command that needs a prompt... This goes back to why can't we just have a delete command

c-ryan-k replied:
Agree, we have talked offline about adding a confirmation to this command if `--clean` is provided.

vilit1 commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,88 @@ def load_iothub_arguments(self, _):
+            "https://github.com/Azure/azure-iot-cli-extension/tree/dev/docs/samples/sample_devices_config.yaml",
+        )
+        context.argument(
+            "device_auth_type",
```
can a person create a sym key device and a cert device in the same command? 

Not sure if we should be explicit here with "Device to hub authorization mechanism for all devices created in this command run" (will need to tweak wording)

c-ryan-k replied:
> can a person create a sym key device and a cert device in the same command?

Not in the same command, no. You can make one run of symmetric key devices, then another run for cert devices, etc - but only one auth type per `devices create` command.

vilit1 commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,88 @@ def load_iothub_arguments(self, _):
+        context.argument(
+            "config_file",
+            options_list=["--config-file", "--config", "--cfg"],
+            help="Path to devices configuration file. Sample configuration file: "
```
are there any specifics for config files with respect to devices? Ex: each device must have a config associated with it

c-ryan-k replied:
This is the main configuration file that contains your YAML or JSON list of devices, think of it as a "device deployment playbook".

vilit1 replied:
I see that if devices and config file are present then the command will error.

vilit1 commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +246,88 @@ def load_iothub_arguments(self, _):
+        )
+        context.argument(
+            "device_config_template",
+            options_list=["--device-template", "--dt"],
```
whats the difference between this and config file?

c-ryan-k replied:
This is the edge `config.toml` device configuration template used as a basis for each edge [device's config](https://learn.microsoft.com/en-us/azure/iot-edge/how-to-provision-single-device-linux-x509?view=iotedge-1.4&tabs=azure-portal%2Cubuntu#provision-the-device-with-its-cloud-identity):

See [device_config.toml](https://github.com/Azure/azure-iot-cli-extension/pull/591/files/db13535a20ae1d5f600b1fd6b67c127d79226cca#diff-7ca392c8d88a5cc372d64ce3e35b9ede99a5ef0f7951edd65f264e9bf4808cf1)

vilit1 commented on setup.py:
```diff
@@ -53,7 +53,10 @@
+    "tomli",
+    "tomli-w",
     "tqdm~=4.62",
+    "treelib",
```
would like to see a pipeline run with 3.7 to make sure everything still works

c-ryan-k replied:
https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=7095&view=logs&j=6b37a07b-ae07-5f7f-46c3-5ce679ba41f2&t=6495c3e0-3a6f-5c64-f017-67dfb92d7cfa

vilit1 commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,426 @@
+        query_args = ["SELECT deviceId FROM devices"]
+        query_method = self.service_sdk.query.get_twins
+        existing_devices = _execute_query(query_args, query_method)
+        existing_device_ids = list(map(lambda x: x["deviceId"], existing_devices))
```
`[x for x["deviceId"] in existing_devices]` does not work?

vilit1 commented on azext_iot/iothub/providers/helpers/edge_device_config.py:
```diff
@@ -0,0 +1,577 @@
+EDGE_LINUX_TUTORIAL_LINK = "https://aka.ms/iotedge-provision-linux-device"
+EDGE_WINDOWS_TUTORIAL_LINK = "https://aka.ms/iotedge-provision-windows"
+
+DEVICE_README = f"""
```
should this be a separate file rather than a constant?

vilit1 replied:
lots of these constants are scripts which make me question if they should be files instead

c-ryan-k replied:
Had a brief conversation about this before you left re: storing files in the extension, and then reading / writing them to device configs, or storing them as a string in code so we can just write to new files without reading the existing ones.

@digimaun curious if you have any thoughts here? I'm unsure of implications of including additional source files for these scripts. It would certainly make them easier to edit and format (and would probably save the linter some struggles) but is that worth the additional read steps every time we need to write these to a device bundle?

digimaun replied:
As we've discussed in the past (at length), in general I think device side configuration scripts on a service side focused client tool (meant to be integrated) is an odd fit, but as this is the path we are experimenting down...

There is convenience organizing all the edge device side elements in a single module. Focusing on the scripts, I see that wrapping scripts as strings makes them easier to use in generating device bundles, but there are trade-offs as you say like formatting, linting and I'm thinking even vulnerability vectors may be harder to detect.

As a data point, we have prior art in picking up non-python code/json schema assets as part of the packaging configuration  (though I think maintaining a python dict representing the schema in code is OK) and they being used for edge deployment validation but

Assuming these scripts are already vetted and used elsewhere, I would re-iterate this changeset is classified as experimental and will be released as such, where things are dynamic and there could be significant changes in functionality/implementation. So in conclusion given the vetted scripts assumption I don't feel strongly on shifting the pattern in this change set.

vilit1 commented on azext_iot/iothub/providers/helpers/edge_device_config.py:
```diff
@@ -0,0 +1,577 @@
+
+    device_toml[
+        "trust_bundle_cert"
+    ] = f"file:///etc/aziot/certificates/{EDGE_ROOT_CERTIFICATE_FILENAME}"
```
is this file path going to always be the same for any system?

c-ryan-k replied:
yes, this is the path the edge agent uses to store certificates and a copy of the device config toml.

vilit1 commented on azext_iot/tests/utility/test_iot_file_operations.py:
```diff
@@ -0,0 +1,59 @@
+            if remove_after_test:
+                os.remove(join(destination, file_name))
+                if destination not in current_or_empty_dirs:
+                    os.rmdir(destination)
```
Add a check that the error did actually get raised and caught in the exception - 
`assert error is None`

c-ryan-k replied:
Will address any gaps in test coverage (by lines etc) in a follow-up PR

vilit1 commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,426 @@
+                rmtree(device_cert_output_directory)
+
+        # Give device registry a chance to catch up
+        sleep(1)
```
this may not be enough - we should have retries with sleeps or have a better error so users wont be confused (and understand that the devices got created but the parent child relationships were not:
![image](https://user-images.githubusercontent.com/73560279/206785977-f9610af6-2cae-497a-8b4a-d41aaed262f7.png)

vilit1 commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,426 @@
+        )
+        for device_id in device_to_parent_iterator:
+            # get device properties
+            device = self.service_sdk.devices.get_identity(id=device_id)
```
why not use existing commands to set device parents?

c-ryan-k replied:
At the time (before I was doing a separate GET for the device identity before the PUT), I already had the entire device object constructed and there was no need to waste queries / throttle limits on extra calls (`set-parent` does a GET for both devices before the PUT on the new device property) - at this point it's probably a wash, though I think the goal with this provider is to eventually move over all of the `hub.py` device-identity commands into it, so eventually this will be able to make use of more provider/class methods

vilit1 commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,426 @@
+            bundle_plural = '' if num_bundles == 1 else 's'
+            print(f"{num_bundles} device bundle{bundle_plural} created in folder: {abspath(bundle_output_directory)}")
+
+    # TODO - Unit test
```
todo

vilit1 commented on azext_iot/tests/iothub/devices/test_iot_edge_devices_create_int.py:
```diff
@@ -0,0 +1,327 @@
+        self._validate_results(devices, "device_bundles", cert_auth=False, custom_device_template=True)
+
+    def _validate_results(self, devices, output_path, cert_auth=False, custom_device_template=False):
+        # device_tuple[0] == device_id
```
im surprised that you didnt use a named tuple here

vilit1 commented on azext_iot/tests/iothub/devices/test_iot_edge_devices_create_unit.py:
```diff
@@ -0,0 +1,1147 @@
+                assert exists(join(output, f"{device_id}.tgz"))
+
+            rmtree(output)
+        pass
```
pass

vilit1 commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,426 @@
+                filter(lambda id: id in device_config_dict, existing_device_ids)
+            )
+            if any(duplicates):
+                raise InvalidArgumentValueError(
```
not sure if there is a test for this yet

c-ryan-k replied:
Have a note to address this in a follow-up PR

c-ryan-k committed:
Wave of PR feedback

Certificate function clarity and docstrings

Utility TOML/YAML error string consolidation

Help and Param updates

c-ryan-k committed:
Added confirmation to '--clean' arg (and '--yes' bypass)

c-ryan-k committed:
Added retry logic to device scope query step

c-ryan-k committed:
Test updates

Added tests for device scope retry logic

Updated int tests to use tuples for device configs

Added unit test for delete_device_identities

c-ryan-k committed:
Merge branch 'dev' into nested_edge_device_bundle

c-ryan-k committed:
Merge branch 'dev' into nested_edge_device_bundle

c-ryan-k committed:
Fix default value for --clean prompt

c-ryan-k committed:
load CA certs with read_file_content instead of open_certificate

c-ryan-k committed:
Merge branch 'dev' into nested_edge_device_bundle

c-ryan-k committed:
Merge branch 'dev' into nested_edge_device_bundle

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -122,3 +132,252 @@ def open_certificate(certificate_path: str) -> str:
+
+
+def create_v3_self_signed_root_certificate(
+    subject: str = "Azure_IoT_CLI_Extension_Cert",
```
Are there consolidation opportunities between this function and other cert creation functions in this module? This function supports v3 extensions while there are more parameters/options for create_self_signed_certificate. There is also the currently named create_ca_signed_certificate below.

If there is some new extension or cert configuration down the line, does that necessitate a new create cert function?

Basically looking to reduce the need of new certificate creation functions in favor of less, but flexible ones with common code if possible.

c-ryan-k replied:
There certainly are.

In developing the v3 extension version, since it was so specifically tailored for edge, I wanted to avoid stepping on existing certificate code until it was finalized. Now that we have it successfully validated for Edge / module certificate communication, there's definitely an opportunity to consolidate.

I'll spike on this for a bit - For the initial self-signed cert, either a way for us to pass extra properties (v3 extensions, etc) into the existing `create_self_signed_certificate` or possibly seeing if we can just apply the v3 extension configuration to existing certificates and make sure that doesn't break anything.

I'll see if that leads to any insight on how to consolidate the signing / cert chaining process for edge certs too

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -122,3 +132,252 @@ def open_certificate(certificate_path: str) -> str:
+    cert_output_dir: Optional[str] = None,
+    cert_file: Optional[str] = None,
+    key_size: int = 4096,
+    valid_days: Optional[int] = 365,
```
Is `valid_days` really Optional? What happens if you pass in `None`?

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -122,3 +132,252 @@ def open_certificate(certificate_path: str) -> str:
+
+def create_ca_signed_certificate(
+    subject: str,
+    ca_public: str,
```
nit, since description states ca_public / ca_private are expecting keys adding _key to the names makes the parameters more self describing.

digimaun commented on azext_iot/common/fileops.py:
```diff
@@ -0,0 +1,47 @@
+from os import makedirs, remove, listdir
```
module is missing headers (coding utf8 (less important for py3 but lets be consistent & explicit) and license.

digimaun reviewed (commented):
Chunk of review

digimaun commented on azext_iot/common/certops.py:
```diff
@@ -122,3 +132,252 @@ def open_certificate(certificate_path: str) -> str:
+    }
+
+
+def create_ca_signed_certificate(
```
It makes sense.

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -316,6 +316,53 @@ def load_iothub_help():
+        long-summary: |
+          This operation accepts inline device arguments or an edge devices configuration file in YAML or JSON format.
+          Inline command args (like '--device-auth') will take precedence and override configuration file properties if they are provided.
+          A sample configuration file can be found here: https://github.com/Azure/azure-iot-cli-extension/tree/dev/docs/samples/sample_devices_config.yaml
```
nit/side thought, should we use an aka.ms link for the sample? 

Unrelated to this PR I want to use an aka.ms link initially pointing to our wiki dedicated page for inline json help for json heavy commands.

c-ryan-k replied:
Yeah, this absolutely should be an aka.ms link - not only can we then update URLs if anything changes (branch/folder structure, filename) without updating help docs, but it's also shorter.

c-ryan-k committed:
Added encoding and license info to fileops.py

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -316,6 +316,53 @@ def load_iothub_help():
+
+    helps["iot edge devices create"] = """
+        type: command
+        short-summary: Create multiple edge devices in an IoT Hub.
```
nit, "Create **and configure**  multiple edge devices in an IoT Hub."

digimaun commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,437 @@
+
+        # Clear devices if necessary
+        if clean and len(existing_device_ids):
+            if not yes and not prompt_y_n(msg=f"Confirm you want to delete all devices in '{self.hub_name}'", default='y'):
```
Since this is destructive, should the default be no/n instead?

c-ryan-k committed:
Consolidated root certificate creation logic

digimaun commented on azext_iot/iothub/commands_device_identity.py:
```diff
@@ -0,0 +1,75 @@
+def iot_delete_devices(
+    cmd,
+    device_ids: List[str],
+    confirm: bool = True,
```
This is not command publicly exposed right? Looking at confirm with default value of `True` here.

c-ryan-k replied:
Oh, that's right. When I was initially creating the method (the internal provider method `delete_device_identities`), instead of `confirm` being used as a way to let the user confirm the deletion, this was an option on the method on whether we should confirm that all devices *were* in fact deleted.

Since it's not publicly exposed (yet) - I don't think this will cause parameter confusion with our other `confirm` options, but I can at least make the parameter in the internal method more explicit and remove it from the provider-facing entrypoint.

digimaun replied:
Understood

digimaun commented on azext_iot/iothub/params.py:
```diff
@@ -249,6 +249,98 @@ def load_iothub_arguments(self, _):
+            "config_file",
+            options_list=["--config-file", "--config", "--cfg"],
+            help="Path to devices configuration file. Sample configuration file: "
+            "https://github.com/Azure/azure-iot-cli-extension/tree/dev/docs/samples/sample_devices_config.yaml",
```
Another spot to replace with aka.ms link

digimaun commented on azext_iot/iothub/params.py:
```diff
@@ -249,6 +249,98 @@ def load_iothub_arguments(self, _):
+        )
+        context.argument(
+            "default_edge_agent",
+            options_list=["--default-edge-agent", "--default-agent", "--dea"],
```
`--dea` :smile:

digimaun commented on azext_iot/iothub/params.py:
```diff
@@ -249,6 +249,98 @@ def load_iothub_arguments(self, _):
+        )
+        context.argument(
+            "device_config_template",
+            options_list=["--device-template", "--dt"],
```
Just an observation that `--device-template`  is somewhat overloaded,  in that "device templates" are used in central [1](https://learn.microsoft.com/en-us/azure/iot-central/core/concepts-device-templates) [2](https://github.com/Azure/azure-iot-cli-extension/issues/364), DTDL (where it transitioned to `model` as the preferred term) and ...

c-ryan-k committed:
Merge branch 'nested_edge_device_bundle' of https://github.com/c-ryan-k/azure-iot-cli-extension into nested_edge_device_bundle

c-ryan-k committed:
PR feedback items

Remnamed ca_cert input vars for create_ca_signed_certificate

Changed default value to 'n' for --clean

Updated help text and params with aka.ms link

Changed --device-template to --device-config-template

digimaun commented on azext_iot/iothub/providers/helpers/edge_device_config.py:
```diff
@@ -0,0 +1,579 @@
+        "trust_bundle_cert"
+    ] = f"file:///etc/aziot/certificates/{EDGE_ROOT_CERTIFICATE_FILENAME}"
+    # Dynamic, AlwaysOnStartup, OnErrorOnly
+    device_toml["auto_reprovisioning_mode"] = "Dynamic"
```
Are AlwaysOnStartup and OnErrorOnly not settable?

c-ryan-k replied:
That was my understanding (at least of how the existing tool functions) after a few sessions with the edge team to debug the resultant config.toml / E2E device deployments - the comments here are just a note of the allowed model values from Edge in Rust. I can connect with them to see if it makes sense to make this a parameterized value, or if we should move this `set` to just the default config template and let an existing TOML file use its own value.

digimaun replied:
Ok gotcha...conceptually (abstract from this particular comment) the more settings we hard code / remove the customization from users the more specificity of scenario this tooling solution becomes. Not sure how important customization of this auto reprovisioning mode setting is.

c-ryan-k replied:
Have a note to follow up in a subsequent PR - I'll change this to:
-  Only set `Dynamic` in our default loaded config, or if there is no value in a supplied user template
-  Check supplied template values to ensure this value is valid
-  Add tests to confirm

digimaun commented on azext_iot/iothub/providers/helpers/edge_device_config.py:
```diff
@@ -0,0 +1,579 @@
+    return config
+
+
+def assemble_nargs_to_dict(hash_list: List[str]) -> Dict[str, str]:
```
If this is exactly the same as the classmethod in ..../deviceupdate/providers/base (looks like it is) we can plug it in common/utility (TBD future reorganization) and reference it from both places

digimaun commented on azext_iot/iothub/providers/device_identity.py:
```diff
@@ -0,0 +1,437 @@
+                raise AzureResponseError(err)
+        if confirm:
+            existing_devices = self.service_sdk.devices.get_devices()
+            if len(existing_devices):
```
minor thing - `list` with an item is True (and empty list is False) so you could do `if existing_devices:`

digimaun commented on azext_iot/tests/utility/test_iot_file_operations.py:
```diff
@@ -0,0 +1,59 @@
+from azext_iot.common.fileops import tar_directory, write_content_to_file
```
Missing headers, I'm thinking we should add a module linter job to help with this kind of thing

digimaun commented on setup.py:
```diff
@@ -53,7 +53,10 @@
     "jsonschema~=3.2.0",
     "importlib_metadata;python_version<'3.8'",
     "azure-iot-device~=2.11",
+    "tomli",
```
It's good practice to constrain dependency so we stay within the major version range we tested with. The concern is a new major version makes a breaking change.

digimaun reviewed (approved):
Nice work and solid tests 

Lets make sure to show evidence of full test pass prior to merge and capture follow ups on our backlog

c-ryan-k committed:
Consolidated multiple instances of assemble_nargs_to_dict to utility method

c-ryan-k committed:
Removed confirm boolean from delete_device_identities

Moved logic into --clean functionality

c-ryan-k committed:
Constrained dependency versions for tomli, tomli-w, treelib

c-ryan-k committed:
Minor PR feedback

c-ryan-k committed:
Fixed another valid_days typing

c-ryan-k committed:
Fixed incorrect format string specifier

c-ryan-k committed:
Configure relative paths for config file values

Makes paths inside config file relative to config file

Does not apply relative pathing to override params (CLI context)

Updates tests and adds nested folder for deployments to test pathing differences

c-ryan-k commented:
First round of tests - have one final one (hopefully) coming that updates unit and integration tests so I'll have another run starting shortly.

https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=7310&view=results

c-ryan-k commented:
Ok, [one last full pipeline run to go](https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=7315&view=results) once the PR pipeline passes, but here's some local testing evidence:
![image](https://user-images.githubusercontent.com/13545962/211444199-ad3adc8e-ce9a-496d-a9ea-04b9d88d5461.png)
![image](https://user-images.githubusercontent.com/13545962/211444225-913e9cb6-d8f7-4de3-bd01-435fdc21a895.png)

c-ryan-k commented:
[Looks like tests were mostly successful](https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=7315&view=ms.vss-test-web.build-test-results-tab), save some of our recent recurring issues. I'm good with this.

c-ryan-k committed:
Add history entry for `edge devices create` command

digimaun reviewed (approved):
LGTM
